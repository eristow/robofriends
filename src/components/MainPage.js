import React, { useEffect } from 'react';

import './MainPage.css';
import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundary from '../components/ErrorBoundary';
import Header from '../components/Header';

const MainPage = ({
  searchField,
  robots,
  isPending,
  error,
  onSearchChange,
  onRequestRobots,
}) => {
  useEffect(() => {
    onRequestRobots();
  }, [onRequestRobots]);

  if (isPending) {
    return <h1>Loading</h1>;
  }

  if (robots.length === 0 || Object.keys(robots).length === 0 || error) {
    return <h1>Error: {error}</h1>;
  }

  const filteredRobots = robots.filter((robot) =>
    robot.name.toLowerCase().includes(searchField.toLowerCase())
  );

  return (
    <div className="tc">
      <Header />
      <SearchBox
        id="searchBox"
        searchChange={onSearchChange}
        searchField={searchField}
      />
      <Scroll>
        <ErrorBoundary>
          <CardList id="cardList" data={filteredRobots} />
        </ErrorBoundary>
      </Scroll>
    </div>
  );
};

export default MainPage;
