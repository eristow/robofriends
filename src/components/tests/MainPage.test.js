import React from 'react';
import { shallow, render, mount } from 'enzyme';
import MainPage from '../MainPage';

describe('MainPage', () => {
  let wrapper;
  let mockProps;

  beforeEach(() => {
    mockProps = {
      onRequestRobots: jest.fn(),
      robots: [
        {
          id: 3,
          name: 'John Joans',
          email: 'joans@gmail.com',
        },
        {
          id: 4,
          name: 'Sarah Joans',
          email: 'sarah@gmail.com',
        },
      ],
      searchField: '',
      isPending: false,
    };
    wrapper = shallow(<MainPage {...mockProps} />);
  });

  it('should render the MainPage component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render loading message when isPending is false', () => {
    mockProps = { ...mockProps, isPending: true };
    wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.text()).toEqual('Loading');
  });

  it('should render error message when error is present', () => {
    const errorMessage = 'an error occurred';
    mockProps = { ...mockProps, error: errorMessage };
    wrapper = shallow(<MainPage {...mockProps} />);
    expect(wrapper.text()).toEqual(`Error: ${errorMessage}`);
  });

  it('should filter robots correctly when a robot exists', () => {
    mockProps = {
      ...mockProps,
      searchField: 'john',
    };
    wrapper = mount(<MainPage {...mockProps} />);
    expect(wrapper.find('#cardList').children().length).toBe(1);
  });

  it('should render no robots when filter returns nothing', () => {
    mockProps = {
      ...mockProps,
      searchField: 'x',
    };
    wrapper = mount(<MainPage {...mockProps} />);
    expect(wrapper.find('#cardList').children().length).toBe(0);
  });
});
