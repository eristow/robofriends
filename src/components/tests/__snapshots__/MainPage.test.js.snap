// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should render the MainPage component 1`] = `
<div
  className="tc"
>
  <Memo() />
  <SearchBox
    id="searchBox"
    searchField=""
  />
  <Scroll>
    <ErrorBoundary>
      <CardList
        data={
          Array [
            Object {
              "email": "joans@gmail.com",
              "id": 3,
              "name": "John Joans",
            },
            Object {
              "email": "sarah@gmail.com",
              "id": 4,
              "name": "Sarah Joans",
            },
          ]
        }
        id="cardList"
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
